pkg/rancher-components/src/components/StringList/StringList.vue

```diff
@@ -0,0 +1,573 @@
+      default: null,
+    },
+    /**
+     * Determines the action buttons position at the bottom of the main box
+     */
+    actionsPosition: {
+      type:    String,
+      default: 'right',
+    },
+    /**
+     * Custom Error messages
+     */
+    messages: {
+      type: Object,
+      default() {
+        return { duplicate: 'Error, item is duplicate!' };
+      },
+    },
+  },
+  data() {
+    return {
+      value:        null as string | null,
+      selected:     null as string | null,
+      isEditItem:   null as string | null,
+      isCreateItem: false,
+      errors:       { duplicate: false } as Record<Error, boolean>,
+    };
+  },
+
+  computed: {
+
+    /**
+     * Create an array of error messages, one for each current error
+     */
+    errorMessages(): string[] {
+      return Object
+        .keys(this.errors)
+        .filter(f => !!(this.errors)[f as Error])
+        .map(k => this.messages[k]);
+    },
+  },
+
+  watch: {
+    /**
+     * Shutdown any user actions on the component when it becomes readonly
+     */
+    readonly() {
+      this.toggleEditMode(false);
+      this.toggleCreateMode(false);
+    },
+  },
+
+  methods: {
+    onChange(value: string) {
+      this.value = value;
+      /**
+       * Remove duplicate error when a new value is typed
+       */
+      this.toggleError(
+        'duplicate',
+        false,
+        this.isCreateItem ? INPUT.create : INPUT.edit,
+      );
+    },
+
+    onSelect(item: string) {
+      if (this.isCreateItem || this.isEditItem === item) {
+        return;
+      }
+      this.selected = item;
+    },
+
+    onSelectNext(arrow: Arrow) {
+      const index = findIndex(
+        this.items as string[],
+        this.selected as string,
+      );
+
+      if (index !== -1) {
+        /**
+         * Select next item in the arrow's direction if it exists,
+         * else select again this.selected (do nothing)
+         */
+        const item = (this.items[index + DIRECTION[arrow]] || this.selected) as string;
+
+        this.onSelect(item);
+        this.moveScrollbar(arrow);
+      }
+    },
+
+    onSelectLeave() {
+      this.selected = null;
+    },
+
+    onClickEmptyBody() {
+      if (!this.isCreateItem && !this.isEditItem) {
+        this.toggleCreateMode(true);
+      }
+    },
+
+    onClickPlusButton() {
+      this.onSelectLeave();
+      this.toggleCreateMode(true);
+    },
+
+    onClickMinusButton() {
+      if (this.isCreateItem) {
+        this.toggleCreateMode(false);
+
+        return;
+      }
+      if (this.isEditItem) {
+        this.toggleEditMode(false);
+
+        return;
+      }
+      if (this.selected) {
+        const index = this.deleteItem(this.selected);
+
+        if (index !== -1) {
+          /**
+           * Select the next item in the list when an item is deleted.
+           * If the deleted item was the last item, select the previous item (this.items[index - 1]). 
+           */
+          setTimeout(() => {
```

Comment:
Oh I missed this timeout. May ask you why? Usually is a sign of issues with some reactivity. Plus you should always unset as garbage collection logic.

Reply:
When an item is removed from the list (`this.items`), the new list is emitted by `change` event, then the list is refreshed by the parent of `StringList`and this could take some times. The selection of the next item in the list cannot happen until the new list is pushed back to the component.

Reply:
`$nextTick` is definitely better and a valid option.
I would however recommend you 2 different solutions for the 2 cases.

For input focus in `setFocus()`, you should rely on autofocus while toggling it with the condition. In the case of the `LabelInput`, it has a prop for that.

For the string selection happening in `onClickMinusButton()`, you could find and assign the new string into `selected` before waiting the DOM gets updated by deletion. Since the 2 operations are not bound with the Node but rather from an UX perspective, we have no constrains. 
You should be able to compute the new index based on the current one. If you are using strings over index, as you may want to have the `selected` value, you would have to map it back and forth to an index.